src/lightning_lite/strategies/xla.py

```diff
@@ -155,20 +155,26 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
         return obj
 
     def all_gather(self, tensor: Tensor, group: Optional[Any] = None, sync_grads: bool = False) -> Tensor:
-        """
-        Function to gather a tensor from several distributed processes
+        """Function to gather a tensor from several distributed processes.
+
         Args:
             tensor: tensor of shape (batch, ...)
             group: not available with TPUs
-            sync_grads: not available with TPUs
+            sync_grads: flag that allows users to synchronize gradients for the all_gather operation
         Return:
             A tensor of shape (world_size, batch, ...)
         """
         if isinstance(tensor, Tensor) and tensor.dim() == 0:
             tensor = tensor.unsqueeze(0)
-        import torch_xla.core.xla_model as xm
 
-        return xm.all_gather(tensor)
+        if sync_grads:
```

Comment:
Is there any disavantage to always using `xf.all_gather`? If no, then let's just switch the function

Reply:
I assume there's a certain communication cost that comes with returning the gradients, even though I cannot say if this actually a problem when running with TPUs.

However, if someone calls the method with `sync_grad=False`, he probably expects that the gradients are not returned, but `xf.all_gather` would still return them. That's why I have the feeling that a switch between the two functions could be better, I saw [this code for the DDP strategy](https://github.com/Lightning-AI/lightning/blob/6f16e46bdb397bf1898e4d7d9050492924177abd/src/lightning_lite/utilities/distributed.py#L201) which is doing it similarly.

Reply:
I expect both functions to be equal when run under a `no_grad` context manager. For example, `torch.distributed`does not offer different implementations whether grads are supported or not.

In fact, looking at the implementation: https://github.com/pytorch/xla/pull/2086/files, it seems like they just copied PyTorch's implementation but using xla's collective ops.

Note that the linked code for `AllGatherGrad` is outdated: #11531

Reply:
Thanks for your explanation, seems like I didn't really understand your initial comment 😅 

Just to make sure that I got it right this time, are you suggesting to do it like in the following snippet? 
```
with nullcontext() if sync_grads else torch.no_grad():
    return xf.all_gather(tensor)
```

Reply:
No, just `xf.all_gather(tensor)`, without the context manager. That will be set elsewhere

Reply:
Sorry, I still don't get it, what do you mean exactly when you say it "will be set elsewhere"?

Do you suggest that if a pytorch-lightning user does not want gradients, he should call it like this:
```
with torch.no_grad():
    self.all_gather(some_tensor)
```
If that is what you meant, I agree it's more straight-forward. The only small problem is that the `sync_grads` parameter won't be taken into account (which may surprise the user). Even though, I have to say that I don't think that it is very intuitive that the default of `sync_grads` is `False` because I assume users prefer to have the gradients (and with your proposal, it would be very easy to achieve.

Having said that I am totally fine to simply use `xf.all_gather(tensor)`, I could update the docstring that explains that the `sync_grads` param is not taken into account to make users aware that the behavior is a bit different from the DDP strategy.

Reply:
> he should call it like this:

The trainer sets the context manager already: https://github.com/Lightning-AI/lightning/blob/24c26f7db298e7fd83e4a37459a46d9043763b0e/src/pytorch_lightning/trainer/trainer.py#L2225-L2238


> I could update the docstring that explains that the sync_grads param is not taken into account to make users aware that the behavior is a bit different from the DDP strategy.

Yes, let's do that!

The parameter is a bit of a legacy one so it's fine to have it not do anything.